File: c++/src/mcmc/timer.cc
#include "mcmc/timer.h"

#include <iomanip>

namespace mcmc {
namespace timer {

Timer::Timer(const std::string &name)
  : name(name), t_total(std::chrono::duration<int>(0)), N(0) {}

void Timer::reset() {
  t_total = std::chrono::duration<int>(0);
  N = 0;
}

void Timer::setTabular(bool on) { tabular = on; }

void Timer::printHeader(std::ostream &s) {
  if (tabular) {
    s << std::left << std::setw(nameWidth + 1) << "timer";
    s << std::right << std::setw(totalWidth + 1) << "total (s)";
    s << std::right << std::setw(tickWidth + 1) << "ticks";
    s << std::right << std::setw(perTickWidth + 1) << "per tick (us)";
    s << std::right << std::endl;
  }
}

std::ostream &Timer::put(std::ostream &s) const {
  using namespace std::chrono;

  s << std::setw(nameWidth) << std::left << name;
  if (N == 0) {
    s << "<unused>";
  } else {
    if (tabular) {
      s << " " << std::setprecision(3) << std::right << std::setw(totalWidth)
        << (duration_cast<milliseconds>(t_total).count() / (double)MILLI);
      s << " " << std::right << std::setw(tickWidth) << N;
      s << " " << std::setprecision(3) << std::fixed << std::right
        << std::setw(perTickWidth)
        << (duration_cast<nanoseconds>(t_total).count() / (double)MILLI / N);
    } else {
      s << " total " << std::setprecision(3) << std::right
        << std::setw(totalWidth)
        << (duration_cast<milliseconds>(t_total).count() / (double)MILLI)
        << "s";
      s << "; ticks " << std::right << std::setw(tickWidth) << N;
      s << "; per tick " << std::fixed << std::setprecision(3) << std::right
        << std::setw(perTickWidth)
        << (duration_cast<nanoseconds>(t_total).count() / (double)MILLI / N)
        << "us";
    }
  }

  return s;
}

std::ostream &operator<<(std::ostream &s, const Timer &timer) {
  return timer.put(s);
}

bool Timer::tabular = false;

}  // namespace timer
}  // namespace mcmc

File: python/com/uva/timer.py
import timeit
import sys

class Timer(object):
    def __init__(self, name):
        self._name = name
        self._N = 0
        self._total = 0

    def start(self):
        self._start = timeit.default_timer()

    def stop(self):
        self._total = timeit.default_timer() - self._start
        self._N += 1

    def print_header(self, out = sys.stdout):
        out.write("%-36s %12s %8s %14s\n" % ("timer", "total (s)", "ticks", "per tick (us)"))

    def report(self, out = sys.stdout):
        out.write("%-36s " % self._name)
        if self._N == 0:
            out.write("<unused>")
        else:
            out.write("%12.3f " % (self._total))
            out.write("%8d " % (self._N))
            out.write("%14.3f" % ((self._total / self._N) * 1000000.0))
        out.write("\n")
        
        
